test/rateLimiters/fixedWindow.test.ts

```diff
@@ -0,0 +1,128 @@
+        describe('after an ALLOWED request...', () => {
+            afterEach(() => {
+                client.flushall();
+            });
+            test('current time window has no token initially', async () => {
+                // zero token used in this time window
+                const withdraw5 = 5;
+                expect((await limiter.processRequest(user1, timestamp, withdraw5)).tokens).toBe(
+                    CAPACITY - withdraw5
+                );
+                const tokenCountFull = await getWindowFromClient(client, user1);
+                expect(tokenCountFull.currentTokens).toBe(5);
+            });
+            test('reached 40% capacity in current time window and still can pass request', async () => {
+                const initial = 5;
+                const partialWithdraw = 2;
+                expect(
+                    (await limiter.processRequest(user2, timestamp, initial + partialWithdraw))
+                        .tokens
+                ).toBe(CAPACITY - initial - partialWithdraw);
+
+                const tokenCountPartial = await getWindowFromClient(client, user2);
+                expect(tokenCountPartial.currentTokens).toBe(initial + partialWithdraw);
+            });
+
+            test('window is partially full and request has no leftover tokens', async () => {
+                const initial = 6;
+                const partialWithdraw = 5;
+                await setTokenCountInClient(client, user2, initial, timestamp);
+                expect(
+                    (await limiter.processRequest(user2, timestamp, partialWithdraw)).success
+                ).toBe(false);
+                expect(
+                    (await limiter.processRequest(user2, timestamp, partialWithdraw)).tokens
+                ).toBe(0);
+            });
+        });
+        describe('after a BLOCKED request...', () => {
+            afterEach(() => {
+                client.flushall();
+            });
+            test('initial request is greater than capacity', async () => {
+                // expect remaining tokens to be 10, b/c the 11 token request should be blocked
+                expect((await limiter.processRequest(user1, timestamp, 11)).success).toBe(false);
+                // expect current tokens in the window to still be 0
+                expect((await getWindowFromClient(client, user1)).currentTokens).toBe(0);
+            });
+            test('window is partially full but not enough time elapsed to reach new window', async () => {
+                const requestedTokens = 9;
+
+                await setTokenCountInClient(client, user2, requestedTokens, timestamp);
+                // expect remaining tokens to be 1, b/c the 2-token-request should be blocked
+                const result = await limiter.processRequest(user2, timestamp + WINDOW_SIZE - 1, 2);
+
+                expect(result.success).toBe(false);
+                expect(result.tokens).toBe(0);
```

Comment:
I would expect `result.tokens` to be`1` here since the request failed and doesn't update the cache.

Reply:
Will change in the revised version.